admin/provisioner/kubernetes.go

```diff
@@ -0,0 +1,303 @@
+	TemplatePaths  *KubernetesTemplatePaths `json:"template_paths"`
+}
+
+type KubernetesTemplatePaths struct {
+	HTTPIngress string `json:"http_ingress"`
+	GRPCIngress string `json:"grpc_ingress"`
+	Service     string `json:"service"`
+	StatefulSet string `json:"statefulset"`
+}
+
+type KubernetesProvisioner struct {
+	Spec      *KubernetesSpec
+	clientset *kubernetes.Clientset
+	templates *template.Template
+}
+
+type TemplateData struct {
+	Image        string
+	ImageTag     string
+	Host         string
+	CPU          int
+	MemoryGB     int
+	StorageBytes int64
+	DataDir      string
+	Slots        int
+	Names        ResourceNames
+	Annotations  map[string]string
+}
+
+type ResourceNames struct {
+	HTTPIngress string
+	GRPCIngress string
+	Service     string
+	StatefulSet string
+}
+
+func NewKubernetes(spec json.RawMessage) (*KubernetesProvisioner, error) {
+	// Parse the Kubernetes provisioner spec
+	ksp := &KubernetesSpec{}
+	err := json.Unmarshal(spec, ksp)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse kubernetes provisioner spec: %w", err)
+	}
+
+	// Build config from kubeconfig file, this will fallback to in-cluster config if no kubeconfig is specified
+	config, err := clientcmd.BuildConfigFromFlags("", ksp.KubeconfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the clientset for the Kubernetes APIs
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the template definitions
+	templates := template.Must(template.New("").ParseFiles(
+		ksp.TemplatePaths.HTTPIngress,
+		ksp.TemplatePaths.GRPCIngress,
+		ksp.TemplatePaths.Service,
+		ksp.TemplatePaths.StatefulSet,
+	))
+
+	return &KubernetesProvisioner{
+		Spec:      ksp,
+		clientset: clientset,
+		templates: templates,
+	}, nil
+}
+
+func (p *KubernetesProvisioner) Provision(ctx context.Context, opts *ProvisionOptions) (*Allocation, error) {
+	// Get Kubernetes resource names
+	names := p.getResourceNames(opts.ProvisionID)
+
+	// Create unique host
+	host := strings.ReplaceAll(p.Spec.Host, "*", opts.ProvisionID)
+
+	// Define template data
+	data := &TemplateData{
+		ImageTag:     opts.RuntimeVersion,
+		Image:        p.Spec.Image,
+		Names:        names,
+		Host:         strings.Split(host, "//")[1], // Remove protocol
+		CPU:          1 * opts.Slots,
+		MemoryGB:     2 * opts.Slots,
+		StorageBytes: 40 * int64(opts.Slots) * int64(datasize.GB),
+		DataDir:      p.Spec.DataDir,
+		Slots:        opts.Slots,
+		Annotations:  opts.Annotations,
+	}
+
+	// Define the structured Kubernetes API resources
+	httpIng := &netv1.Ingress{}
+	grpcIng := &netv1.Ingress{}
+	svc := &apiv1.Service{}
+	sts := &appsv1.StatefulSet{}
+
+	// Resolve the templates and decode into Kubernetes API resources
+	for k, v := range map[string]any{
+		p.Spec.TemplatePaths.HTTPIngress: httpIng,
+		p.Spec.TemplatePaths.GRPCIngress: grpcIng,
+		p.Spec.TemplatePaths.Service:     svc,
+		p.Spec.TemplatePaths.StatefulSet: sts,
+	} {
```

Comment:
I'm wondering if we could avoid the external templated files and instead just define the base resources directly in code, and override the relevant properties here. Or are the specs pretty internal/custom?

Reply:
I thought quite a bit about this when deciding on the templates and I totally understand the desire to have the resources directly defined in code, but there's just too many environment specific custom configurations of the Kubernetes resources needed for it to really be a viable option.

Reply:
I've added a `README.md` with some initial config instructions and some simplified template examples, in my opinion this should be enough for now and internal Rill people can refer to our private infra repos for our internal configurations and templates. Creating a simple developer workflow for the Kubernetes part of the provisioner is hard and it's something we should tackle as a separate task if we feel it's worth the effort to support.